### PR TITLE
[Scan] Equality delete bound-based pruning based on data file bounds

### PR DESCRIPTION
**Pull request by Tishj**
Implements the optimization proposed and written in https://github.com/duckdb/duckdb-iceberg/issues/341#issuecomment-5009519318

### Summary

Per field of the `equality_ids` of the equality delete file, we determine that we can skip the equality delete if the following are true:
- delete file: null counts for the field_id are set and they are 0.
- delete file: lower+upper bounds present and non-null.
- data file: lower+upper bounds present and non-null.
- if column type is FLOAT or DOUBLE:
  - delete file: nan counts for the field_id are set and they are 0.
- delete.upper_bound < data.lower_bound OR delete.lower_bound > data.upper_bound

### Future Work

Since we have the columns emitted by the data file at this point, we can probably make this pruning even more thorough, if we have bounds information for the delete file, and the data file doesn't have the field id, we can figure out what the constant value for that field-id is, and check that against the delete bounds instead.
Also we can extend this to cover NULLs as well, currently the optimization only prunes for non-null delete files.

### CODE REVIEW

File: src/planning/pruning/iceberg_file_pruner.cpp [hunk 1]
```diff
@@ -245,6 +245,71 @@ bool IcebergFilePruner::DeleteManifestMatchesDataFile(const IcebergManifestFile
...
 }
 
+bool IcebergFilePruner::EqualityDeleteMatchesDataFile(const IcebergDataFile &delete_file,
+                                                      const IcebergDataFile &data_file) const {
+	auto &equality_ids = delete_file.equality_ids;
+
+	for (auto field_id : equality_ids) {
+		auto delete_null_count = delete_file.null_value_counts.find(field_id);
+		if (delete_null_count == delete_file.null_value_counts.end() || delete_null_count->second != 0) {
+			//! A NULL delete key can match NULL data values - require a known zero null count
```
Comment (Tmonster) [inline src/planning/pruning/iceberg_file_pruner.cpp:255]: If the equality delete file is just one `NULL` value the equality delete file will automatically match it seems?

File: src/planning/pruning/iceberg_file_pruner.cpp [hunk 2]
```diff
@@ -245,6 +245,71 @@ bool IcebergFilePruner::DeleteManifestMatchesDataFile(const IcebergManifestFile
...
+			continue;
+		}
+
+		auto delete_lower = delete_file.lower_bounds.find(field_id);
+		auto delete_upper = delete_file.upper_bounds.find(field_id);
+		auto data_lower = data_file.lower_bounds.find(field_id);
+		auto data_upper = data_file.upper_bounds.find(field_id);
+		if (delete_lower == delete_file.lower_bounds.end() || delete_upper == delete_file.upper_bounds.end() ||
+		    data_lower == data_file.lower_bounds.end() || data_upper == data_file.upper_bounds.end()) {
+			continue;
```
Comment (Tmonster) [inline src/planning/pruning/iceberg_file_pruner.cpp:265]: What happens in the following case?

- We start a table with 3 columns, field_ids 1, 2, 3. & insert data
- We add a fourth column, field id 4 & insert data
- We apply an equality delete  only to field_id 4
- We scan.

I feel like this check will hit `continue` since none of the delete file field_ids are in the data file, and we will then try to apply the delete file to a data file that doesn't even have the field id?

File: src/planning/pruning/iceberg_file_pruner.cpp [hunk 3]
```diff
@@ -245,6 +245,71 @@ bool IcebergFilePruner::DeleteManifestMatchesDataFile(const IcebergManifestFile
...
+				DUCKDB_LOG(context, IcebergLogType,
+				           "Iceberg Equality Delete Pruning, skipped 'equality_delete_file': '%s' for 'data_file': "
+				           "'%s', equality field '%s' (field id %d) has bounds [%s, %s] outside data bounds [%s, %s]",
+				           delete_file.file_path, data_file.file_path, column.name, field_id,
+				           delete_stats.lower_bound->ToString(), delete_stats.upper_bound->ToString(),
+				           data_stats.lower_bound->ToString(), data_stats.upper_bound->ToString());
+				return false;
+			}
+		} catch (Exception &) {
+			continue;
```
Comment (Tmonster) [inline src/planning/pruning/iceberg_file_pruner.cpp:312]: Can we add a `DUCKDB_LOG_DEBUG` here?

File: src/planning/pruning/iceberg_file_pruner.cpp (hunk 2 above)
Comment (Tmonster) [reply src/planning/pruning/iceberg_file_pruner.cpp:265]: Or maybe this check already happens earlier?

**Review by Tmonster — COMMENTED** [review]
Thanks! 1 or 2 questions on edge cases

Can we also get a regression test on here? I think We should defs be able to see some performance improvements in certain cases.

File: src/planning/pruning/iceberg_file_pruner.cpp (hunk 2 above)
Comment (Tishj) [reply src/planning/pruning/iceberg_file_pruner.cpp:265]: That's indeed an interesting question, and one that's covered by the spec (https://iceberg.apache.org/spec/#equality-delete-files)

> If a column was added to a table and later used as a delete column in an equality delete file, the column value is read for older data files using normal projection rules (defaults to null)

So no, the delete still applies even though the file doesn't contain the field id

File: src/planning/pruning/iceberg_file_pruner.cpp (hunk 1 above)
Comment (Tishj) [reply src/planning/pruning/iceberg_file_pruner.cpp:255]: Yea that's a correct observation, I'm not entirely sure why, it was just copied from the original issue.
I think we could make this optimization work even if there are NULLs.
